clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -577,33 +582,45 @@ private ListOffsetResult fetchOffsetsByTimes(Map<TopicPartition, Long> timestamp
      */
     public Map<TopicPartition, List<ConsumerRecord<K, V>>> fetchedRecords() {
```

Comment:
Just one more thing I thought of when I was looking at the test cases. With paused partitions in `completedFetches`, I think we will never block in `KafkaConsumer.pollForFetches` since `hasCompletedFetches` would continue returning true even if there is no data that can be returned. This would result in a busy loop consuming cpu. We could either change the logic to `hasCompletedFetches` or we could introduce a new api (e.g. `hasAvailableFetches`).

Reply:
I'll implement `hasAvailableFetches` to check if a fetchable partition exists in `completedFetches`.  `hasCompletedFetches` only has the single call site in the `KafkaConsumer`.  We could make it make it package protected if we wanted to so tests can still use it.  Since this is part of clients internals should we do that?

Reply:
Yeah, that sounds fine to me. It does seem useful for testing, so package access makes sense.